llvmPackages: use libcxxrt on FreeBSD

alyssais opened the pull request:
###### Description of changes

FreeBSD doesn't use LLVM's cxxabi implementation, for backwards compatibility reasons.  Software expects the libcxxrt API when building on FreeBSD.  This fixes the build of `pkgsCross.x86_64-freebsd.boost`.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

alyssais committed:
libcxxrt: init at unstable-2022-08-08

alyssais committed:
llvmPackages: use libcxxrt on FreeBSD

FreeBSD doesn't use LLVM's cxxabi implementation, for backwards
compatibility reasons.  Software expects the libcxxrt API when
building on FreeBSD.  This fixes the build of
pkgsCross.x86_64-freebsd.boost.

alyssais commented:
Eval is failing because of LLVMs 5 and 6 again. >:(

Ericson2314 commented on pkgs/development/compilers/llvm/11/default.nix:
```diff
@@ -125,7 +125,7 @@ let
       libcxx = targetLlvmLibraries.libcxx;
       extraPackages = [
-        targetLlvmLibraries.libcxxabi
+        libcxx.cxxabiLib
```
I think you want
```suggestion
        targetLlvmLibraries.libcxx.cxxabiLib
```
?

alyssais replied:
They're the same thing because of the `rec`, but I'll happily change it to be more explicit if you think it'd be better.

alyssais commented:
Huh, I guess this doesn't really need to target staging.

alyssais committed:
llvmPackages: don't exclude FreeBSD from -lunwind

The exception for FreeBSD was added in 0afe9d1f704 ("freebsd packages:
Init at 13.1"), but it seems to have been erroneous, as e.g. ncurses
fails to build:

	x86_64-unknown-freebsd13-clang++  -o demo ../obj_s/demo.o -L../lib -lncurses++w -L../lib -lformw -lmenuw -lpanelw -lncursesw   -lutil     -DHAVE_CONFIG_H -DBUILDING_NCURSES_CXX -I../c++ -I. -I../include  -DNDEBUG -O2  -fPIC -DPIC
	x86_64-unknown-freebsd13-ld: error: undefined symbol: _Unwind_Resume
	>>> referenced by demo.cc
	>>>               ../obj_s/demo.o:(TestApplication::run())
	>>> referenced by demo.cc
	>>>               ../obj_s/demo.o:(_GLOBAL__sub_I_demo.cc)
	>>> referenced by demo.cc
	>>>               ../obj_s/demo.o:(NCursesUserItem<UserData>::NCursesUserItem(char const*, char const*, UserData const*))
	>>> referenced 46 more times
	clang-12: error: linker command failed with exit code 1 (use -v to see invocation)

This is fixed by undoing the change, adding -lunwind on FreeBSD.

alyssais commented:
I've split the `-lunwind` change into its own commit, since it turns out it's needed independently of the change to libcxxrt.  But I don't think there's any point in splitting it into a separate PR.

Ericson2314 commented on pkgs/build-support/cc-wrapper/default.nix:
```diff
@@ -371,11 +371,11 @@ stdenv.mkDerivation {
       echo "-stdlib=libc++" >> $out/nix-support/libcxx-ldflags
-      echo "-lc++abi" >> $out/nix-support/libcxx-ldflags
-    '')
+      echo "-l${{ libcxxabi = "c++abi"; libcxxrt = "cxxrt"; }.${libcxx.cxxabi}}" >> $out/nix-support/libcxx-ldflags
```
```suggestion
      echo "-l${libcxx.cxxabiLib.libName}" >> $out/nix-support/libcxx-ldflags
```
Can we do something like that?

alyssais replied:
I realised that, apart from this, nothing else was using the string `libcxx.cxxabi`, so I got rid of it, and then renamed what was previously `libcxx.cxxabiLib` to `libcxx.cxxabi`, so now we can do `libcxx.cxxabi.libName` (as `libcxx.cxxabi` is the package now, not a string).

alyssais commented:
Updated to be less stringy based on a suggestion from @Ericson2314 on Matrix.  We now pass `cxxabi` around as a package, and look at its `pname` when necessary to tell them apart, rather than passing a string around and trying to map it back to a packag.
